src/cpp/fastdds/publisher/DataWriterImpl.cpp

```diff
@@ -1564,11 +1573,85 @@ bool DataWriterImpl::deadline_timer_reschedule()
         return false;
     }
 
+    if (next_deadline_us <= std::chrono::steady_clock::now())
+    {
+        // Ignore uninitialized or stale deadlines; don't arm to avoid a spurious first miss
+        return false;
+    }
+
     auto interval_ms = duration_cast<milliseconds>(next_deadline_us - steady_clock::now());
     deadline_timer_->update_interval_millisec(static_cast<double>(interval_ms.count()));
     return true;
 }
 
+void DataWriterImpl::configure_deadline_timer_()
+{
+    std::unique_lock<RecursiveTimedMutex> lock(writer_->getMutex());
+
+    // Create the timer once 
+    if (deadline_timer_ == nullptr)
+    {
+        deadline_timer_ = new TimedEvent(
+            publisher_->rtps_participant()->get_resource_event(),
+            [this]() -> bool
+            {
+                std::unique_lock<RecursiveTimedMutex> lock(writer_->getMutex());
+
+                // Normal path: process a deadline miss. This function will
+                // update counters, notify, and ask for reschedule as needed.
+                return deadline_missed();
+            },
+            (qos_.deadline().period == dds::c_TimeInfinite) ? std::numeric_limits<double>::max() : qos_.deadline().period.to_ns() * 1e-6
+            // In case of deadline period = infinite, multiplying it by 1e-6 could cause a problem
+        );
+    }
+
+    // Handle "infinite" and "zero" outside the callback
+    if (qos_.deadline().period == dds::c_TimeInfinite)
+    {
+        deadline_timer_->cancel_timer();
+        return;
+    }
+
+    deadline_duration_us_ = std::chrono::duration<double, std::ratio<1, 1000000>>(qos_.deadline().period.to_ns() * 1e-3);
```

Comment:
Maybe it's better to keep the maximum value when infinite, as it was the case before.